fix: @Replaces with qualifier doesn't work (fix #7261)

ivyazmitinov opened the pull request:
fix #7261

ivyazmitinov committed:
fix: @Replaces with qualifier doesn't work (fix #7261)

(cherry picked from commit a45e87993ade7255f6fecbcd7755698bcfe64fe9)

ivyazmitinov commented:
@dstepanov, Fixed

ivyazmitinov commented:
I've also got a bit of the refactoring fever and

- Changed `Replaces#qualifier` type to the `Class<? extends Annotation> instead of an array. It required a bit of a workaround, but it is a lesser evil, IMO
- Added exception when both `named` and `qualifier` present 

Both changes are meant to streamline the API

ivyazmitinov committed:
Return logging on factory replacement #7261

ivyazmitinov committed:
Fix tests #7261

ivyazmitinov committed:
Fix checkstyle #7261

ivyazmitinov committed:
Fix accepted-api-changes

ivyazmitinov commented:
@graemerocher, I am a bit confused regarding tests: there is a `io.micronaut.docs.replaces.ReplacesTest`, which I've updated with the new use-cases, but there are also a bunch of `io.micronaut.inject.qualifiers.replaces.ReplacesSpec`s. Should I add tests to them as well, or what's best to do?

graemerocher commented:
the first one is only for the documentation, the last one is where new and existing tests should be updated yes

ivyazmitinov committed:
Add tests #7261

ivyazmitinov commented:
@graemerocher Done :)

dstepanov commented on inject/src/main/java/io/micronaut/context/DefaultBeanContext.java:
```diff
@@ -2319,74 +2322,93 @@ private <T> void filterReplacedBeans(BeanResolutionContext resolutionContext, Co
-                            }
-                        }
-                    }
+            if (named.isPresent() || qualifier.isPresent()) {
```
I would prefer not to have so many `if else` statements, when a block returns the value you don't need to have `else`. Also, I would extract the values only when they are needed:
```
final Optional<String> named = ...
LOG ...
if (named...

Optional<Class<?>> factory
LOG...
if (factory.isPresent() 
etc
```

dstepanov commented on inject/src/main/java/io/micronaut/context/annotation/Replaces.java:
```diff
@@ -54,12 +56,19 @@
      * @return The qualifier
      */
-    Class<? extends Annotation>[] qualifier() default {};
+    Class<? extends Annotation> qualifier() default ReplacesQualifierDefaultValue.class;
```
You can use simple `Annotation.class`

dstepanov commented on inject/src/main/java/io/micronaut/context/DefaultBeanContext.java:
```diff
@@ -2319,74 +2322,93 @@ private <T> void filterReplacedBeans(BeanResolutionContext resolutionContext, Co
-                        }
-                    }
+            if (named.isPresent() || qualifier.isPresent()) {
+                if (named.isPresent() && qualifier.isPresent()) {
```
Here you check if optional is present and then you continue with optional in your code. Please first check if `named` qualifies then if `qualifier` qualifies etc. Preferably reduce the number of lamdas.

ivyazmitinov committed:
Fixes after review #7261

ivyazmitinov commented:
@dstepanov, Fixed all the issues. 
Also, improved the logging a bit: added additional logs when bean is replaced by any qualifier, and now it writes the log only when a bean was actually replaced.

ivyazmitinov commented:
@dstepanov, kind reminder :)

dstepanov commented:
@ivyazmitinov Thanks it looks good! One more thing I noticed is that you can replace the stream with a simple loop in this code:

```java
return replacementTypes.stream()
            .filter(isNotTheSameDefinition(definitionToBeReplaced))
            .filter(isNotProxy(definitionToBeReplaced))
            .anyMatch(checkIfReplaces(definitionToBeReplaced, annotationMetadata));
```
We try to use streams as minimal as possible in the core.

Also, now we have the Sonar which unfortunately doesn't run for contributions but it likes to complain about the code style:
```java
if (...) {
 return ...
} else { // not needed else
 return or throw
}
````
It would be nice if you can remove those elses.

ivyazmitinov commented:
@dstepanov 
Thanks, will do :) 
And I intentionally refactored the original loop into the stream with fancy predicates, wan't aware about "no streams" policy :) 
What is the issue with them, if you don't mind me asking?

dstepanov commented:
We try to not have them in hot paths like a startup (JVM take some time to optimize them) or request processing to reduce allocations.

ivyazmitinov committed:
More fixes after review #7261

ivyazmitinov commented:
@dstepanov, should be ok now

ivyazmitinov committed:
Checkstyle fix

ivyazmitinov commented:
@sdelamo, could you, by any chance, take a look at the PR?

sdelamo commented:
@ivyazmitinov can you merge 3.5.x into the branch and address the test failures?

ivyazmitinov commented:
@sdelamo, done. GraalVM CE CI seems to be unrelated to my changes
